### PR TITLE
BarChart: color by field, x time field, bar radius, label skipping

### LINKED ISSUE #870: Support for multiple series & bars (side by side) for same time point
I have multiple data source on one single graph. For example I have 6 metrics on one single graph.Under chart options when i select bars they all show up on single bar line I want to show them as different bars ? How do i do that ? I have attached a sample image file. 

The problem with having all of them on one single line is the last data source that i add takes up all the space on the bar.

git2

git

images

### PR DESCRIPTION
**Pull request by leeoniya**
Part 3 of 3 #41510 (we decided to break it up to backport selectively)

image

this implements the main chunk of the work:

1. adds a color by field option
2. enables barcharts to use a time field as the x labels. (previously only string labels were accepted)
3. adds a bar radius option for pill-style bars
4. adds a slider to do forwards and backwards density-based x label skipping so it's possible to render e.g. 250 bars without label run-in. currently the control is manual since it's difficult to do all the right "auto" things while accounting for label rotation, max label lengths, user preference, etc. Related: #37966.
5. Tooltip includes all values including ones that are not in the graph.  (additional string fields etc)

Since this PR adds support for x time field in grouped BarChart, lets's finally close this one:

Fixes #870

### CODE REVIEW

**Review by ryantxu — APPROVED** [review]
Looks great!

The only bit I am not sure on is the labels for:
image

Maybe "LTR" and "RTL" -- more cryptic but shorter?  "From right" or "From left"?

**Comment by ryantxu** [comment]
Lets leave labels as is for now and refine after some more testing and socialization :)

**Comment by torkelo** [comment]
This is really nice! the label skipping is a bit complicated to understand at first. And It's confusing how this panel handles time x axis so differently than time series (thinking from a user point of view) requiring manual format & label skipping settings to create x-axis that is readable.  Is there a way we can have time axis handled the same way?

**Comment by zoltanbedi** [comment]
Awesome work! This also fixes #40865
On thing that I think people are going to complain about is that the colors from the tooltip is missing.

Screenshot 2021-12-20 at 9 45 52

Also it seems like that the legend ordering is broken.

https://user-images.githubusercontent.com/13729989/146739106-1bdcb551-4ab6-4c02-939c-ccbed102b8b9.mp4

**Comment by dprokop** [comment]
@leeoniya I've noticed I think wrong behaviour when Color by field option is set, please see the attached dashboard:

image


There are 3 bar groups, 3-values each, but when color by field is chosen only single bar is rendered per group.

<details>
<summary>dashboard.json</summary>
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 638,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "palette-classic"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "axisSoftMin": 0,
            "fillOpacity": 80,
            "gradientMode": "none",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineWidth": 1,
            "scaleDistribution": {
              "type": "linear"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "barRadius": 0,
        "barWidth": 0.97,
        "colorByField": "b",
        "groupWidth": 0.7,
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "orientation": "auto",
        "showValue": "auto",
        "stacking": "none",
        "tooltip": {
          "mode": "single"
        },
        "xTickLabelRotation": 0,
        "xTickLabelSpacing": 0
      },
      "targets": [
        {
          "csvContent": "a,b,c,d\n'a',2,3,4\n'b',5,1,4\n'c',17,3,12",
          "refId": "A",
          "scenarioId": "csv_content"
        }
      ],
      "title": "Panel Title",
      "type": "barchart"
    }
  ],
  "schemaVersion": 34,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Bar chart improvements",
  "uid": "EdHNh-o7k",
  "version": 3,
  "weekStart": ""
}
```
</details>

**Comment by dprokop** [comment]
> Looks great!
> 
> The only bit I am not sure on is the labels for: image
> 
> Maybe "LTR" and "RTL" -- more cryptic but shorter? "From right" or "From left"?

I agree on this one. When I tested this out I was slightly confused about what this option is even supposed to do. Think it would be easier to grasp the idea of this if we used predefined t-shirt sizing (radio button group) and assume LTR direction to start with. What I think about is sth like this:
image

**Comment by ryantxu** [comment]
@dprokop -- I think the issue you see is that field "b" is just using constant color from classic pallet (in this case yellow) if you pick a color scheme that depends on the values, it makes more sense:

image

maybe a warning in the UI would help :thinking:

File: public/app/plugins/panel/barchart/utils.ts [hunk 1]
```diff
@@ -154,7 +170,7 @@ export const preparePlotConfigBuilder: UPlotConfigPrepFn<BarChartOptions> = ({
       theme,
       colorMode,
       pathBuilder: config.barsBuilder,
-      show: !customConfig.hideFrom?.viz,
+      show: !customConfig.hideFrom?.viz, // Not necessary
```
Comment (ryantxu) [inline public/app/plugins/panel/barchart/utils.ts:173]: ^^??

**Comment by ryantxu** [comment]
@dprokop + @leeoniya -- lets remove the `xTickLabelSpacing` config UI from this PR and add it to another.  This one has already gotten too big :grimacing:

**Pull request merged**